<h1>Ишенкалыев Тенгиз AIN-2-22</h1>

<h2>Импорт необходимых библиотек</h2>


In [21]:
from sklearn.datasets import load_diabetes, load_linnerud
from sklearn.model_selection import train_test_split, GridSearchCV
from sklearn.neighbors import KNeighborsRegressor
from sklearn.linear_model import LinearRegression, Ridge, Lasso
from sklearn.metrics import mean_squared_error, r2_score
import pandas as pd


<h2>Загрузка и преобразование датасетов</h2>


In [22]:
diabetes_data = load_diabetes(as_frame=True)
linnerud_data = load_linnerud(as_frame=True)
diabetes_df = pd.concat([diabetes_data['data'], diabetes_data['target']], axis=1)
linnerud_df = pd.concat([linnerud_data['data'], linnerud_data['target']], axis=1)

print("Diabetes Dataset:")
display(diabetes_df.head())

print("Linnerud Dataset:")
display(linnerud_df.head())


Diabetes Dataset:


age       sex       bmi        bp        s1        s2        s3  \
0  0.038076  0.050680  0.061696  0.021872 -0.044223 -0.034821 -0.043401   
1 -0.001882 -0.044642 -0.051474 -0.026328 -0.008449 -0.019163  0.074412   
2  0.085299  0.050680  0.044451 -0.005670 -0.045599 -0.034194 -0.032356   
3 -0.089063 -0.044642 -0.011595 -0.036656  0.012191  0.024991 -0.036038   
4  0.005383 -0.044642 -0.036385  0.021872  0.003935  0.015596  0.008142   

         s4        s5        s6  target  
0 -0.002592  0.019907 -0.017646   151.0  
1 -0.039493 -0.068332 -0.092204    75.0  
2 -0.002592  0.002861 -0.025930   141.0  
3  0.034309  0.022688 -0.009362   206.0  
4 -0.002592 -0.031988 -0.046641   135.0

Linnerud Dataset:


Chins  Situps  Jumps  Weight  Waist  Pulse
0    5.0   162.0   60.0   191.0   36.0   50.0
1    2.0   110.0   60.0   189.0   37.0   52.0
2   12.0   101.0  101.0   193.0   38.0   58.0
3   12.0   105.0   37.0   162.0   35.0   62.0
4   13.0   155.0   58.0   189.0   35.0   46.0

<h2>Описание и проверка пропущенных значений</h2>


In [23]:
print("Описание датасета Diabetes:")
display(diabetes_df.describe())

print("Описание датасета Linnerud:")
display(linnerud_df.describe())

print("Проверка пропущенных значений...")
print("Missing values in Diabetes Dataset:", diabetes_df.isnull().sum().sum())
print("Missing values in Linnerud Dataset:", linnerud_df.isnull().sum().sum())


Описание датасета Diabetes:


age           sex           bmi            bp            s1  \
count  4.420000e+02  4.420000e+02  4.420000e+02  4.420000e+02  4.420000e+02   
mean  -2.511817e-19  1.230790e-17 -2.245564e-16 -4.797570e-17 -1.381499e-17   
std    4.761905e-02  4.761905e-02  4.761905e-02  4.761905e-02  4.761905e-02   
min   -1.072256e-01 -4.464164e-02 -9.027530e-02 -1.123988e-01 -1.267807e-01   
25%   -3.729927e-02 -4.464164e-02 -3.422907e-02 -3.665608e-02 -3.424784e-02   
50%    5.383060e-03 -4.464164e-02 -7.283766e-03 -5.670422e-03 -4.320866e-03   
75%    3.807591e-02  5.068012e-02  3.124802e-02  3.564379e-02  2.835801e-02   
max    1.107267e-01  5.068012e-02  1.705552e-01  1.320436e-01  1.539137e-01   

                 s2            s3            s4            s5            s6  \
count  4.420000e+02  4.420000e+02  4.420000e+02  4.420000e+02  4.420000e+02   
mean   3.918434e-17 -5.777179e-18 -9.042540e-18  9.293722e-17  1.130318e-17   
std    4.761905e-02  4.761905e-02  4.761905e-02  4.761905e-02  4.761905e-02   
min   -1.156131e-01 -1.023071e-01 -7.639450e-02 -1.260971e-01 -1.377672e-01   
25%   -3.035840e-02 -3.511716e-02 -3.949338e-02 -3.324559e-02 -3.317903e-02   
50%   -3.819065e-03 -6.584468e-03 -2.592262e-03 -1.947171e-03 -1.077698e-03   
75%    2.984439e-02  2.931150e-02  3.430886e-02  3.243232e-02  2.791705e-02   
max    1.987880e-01  1.811791e-01  1.852344e-01  1.335973e-01  1.356118e-01   

           target  
count  442.000000  
mean   152.133484  
std     77.093005  
min     25.000000  
25%     87.000000  
50%    140.500000  
75%    211.500000  
max    346.000000

Описание датасета Linnerud:


Chins      Situps      Jumps      Weight      Waist      Pulse
count  20.000000   20.000000   20.00000   20.000000  20.000000  20.000000
mean    9.450000  145.550000   70.30000  178.600000  35.400000  56.100000
std     5.286278   62.566575   51.27747   24.690505   3.201973   7.210373
min     1.000000   50.000000   25.00000  138.000000  31.000000  46.000000
25%     4.750000  101.000000   39.50000  160.750000  33.000000  51.500000
50%    11.500000  122.500000   54.00000  176.000000  35.000000  55.000000
75%    13.250000  210.000000   85.25000  191.500000  37.000000  60.500000
max    17.000000  251.000000  250.00000  247.000000  46.000000  74.000000

Проверка пропущенных значений...
Missing values in Diabetes Dataset: 0
Missing values in Linnerud Dataset: 0


<h2>Разделение данных на тренировочные и тестовые выборки</h2>


In [24]:
X_train_d, X_test_d, y_train_d, y_test_d = train_test_split(
    diabetes_df.drop(columns='target'), diabetes_df['target'], test_size=0.2, random_state=42)

X_train_l, X_test_l, y_train_l, y_test_l = train_test_split(
    linnerud_df.drop(columns=['Weight', 'Waist', 'Pulse']), linnerud_df['Weight'], test_size=0.2, random_state=42)


<h2>Определение моделей и параметров для GridSearchCV</h2>


In [25]:
models = {
    "KNeighborsRegressor": KNeighborsRegressor(),
    "LinearRegression": LinearRegression(),
    "Ridge": Ridge(),
    "Lasso": Lasso()
}

parameters = {
    "KNeighborsRegressor": {"n_neighbors": [3, 5, 7]},
    "Ridge": {"alpha": [0.1, 1.0, 10.0]},
    "Lasso": {"alpha": [0.1, 1.0, 10.0]}
}


<h2>Функция для обучения и оценки модели</h2>


In [26]:
def train_and_evaluate(model, X_train, X_test, y_train, y_test):
    model.fit(X_train, y_train)
    predictions = model.predict(X_test)
    mse = mean_squared_error(y_test, predictions)
    r2 = r2_score(y_test, predictions)
    return mse, r2


<h2>Оценка моделей на датасете Diabetes</h2>


In [27]:
print("Оценка моделей на датасете Diabetes...")
results = []
for model_name, model in models.items():
    if model_name in parameters:
        grid = GridSearchCV(model, parameters[model_name], scoring='r2', cv=5)
        grid.fit(X_train_d, y_train_d)
        best_model = grid.best_estimator_
    else:
        best_model = model
    mse, r2 = train_and_evaluate(best_model, X_train_d, X_test_d, y_train_d, y_test_d)
    results.append((model_name, "Diabetes", mse, r2))


Оценка моделей на датасете Diabetes...


<h2>Оценка моделей на датасете Linnerud</h2>


In [28]:
print("Оценка моделей на датасете Linnerud...")
for model_name, model in models.items():
    if model_name in parameters:
        grid = GridSearchCV(model, parameters[model_name], scoring='r2', cv=5)
        grid.fit(X_train_l, y_train_l)
        best_model = grid.best_estimator_
    else:
        best_model = model
    mse, r2 = train_and_evaluate(best_model, X_train_l, X_test_l, y_train_l, y_test_l)
    results.append((model_name, "Linnerud", mse, r2))


Оценка моделей на датасете Linnerud...


<h2>Сравнение результатов моделей для каждого датасета</h2>


In [29]:
results_df = pd.DataFrame(results, columns=["Model", "Dataset", "MSE", "R2 Score"])
display(results_df)


Model   Dataset          MSE  R2 Score
0  KNeighborsRegressor  Diabetes  2986.975006  0.436223
1     LinearRegression  Diabetes  2900.193628  0.452603
2                Ridge  Diabetes  2856.486888  0.460852
3                Lasso  Diabetes  2798.193485  0.471855
4  KNeighborsRegressor  Linnerud   435.561224 -0.549006
5     LinearRegression  Linnerud   630.775710 -1.243257
6                Ridge  Linnerud   626.295358 -1.227323
7                Lasso  Linnerud   548.542969 -0.950809

<h2>Выбор лучших моделей для каждого датасета</h2>


In [30]:
best_models = results_df.sort_values(by=["Dataset", "R2 Score"], ascending=[True, False])
print("Лучшие модели для каждого датасета:")
display(best_models.groupby("Dataset").first())


Лучшие модели для каждого датасета:


Model          MSE  R2 Score
Dataset                                             
Diabetes                Lasso  2798.193485  0.471855
Linnerud  KNeighborsRegressor   435.561224 -0.549006

<h3> Выводы по работе</h3>

Для датасета Diabetes: Модель Lasso показала наилучшие значения метрик с MSE = 2798.19 и R2 Score = 0.47

Для датасета Linnerud: Модель KNeighborsRegressor оказалась лучшей с MSE = 435.561224, но при этом R2 Score оказался отрицательным (-0.54), что указывает на низкую предсказательную способность модели для этого набора данных

В целом, использование регуляризации и оптимизация гиперпараметров с помощью GridSearchCV помогли улучшить качество предсказаний, а метрики MSE и R2 Score позволили объективно оценить производительность каждой модели.